### PR TITLE
SNOW-945768: Addressing Banned functions in ODBC driver

### PR DESCRIPTION
**Pull request by sfc-gh-ext-simba-jz**
Microsoft made a list of banned CRT functions in their recent review feedback. We should use "safe" version (such as fopen_s for fopen) that they accept.

### CODE REVIEW

File: cpp/FileTransferAgent.cpp [hunk 1]
```diff
@@ -616,12 +617,14 @@ void Snowflake::Client::FileTransferAgent::compressSourceFile(
   stagingFile += m_stmtPutGet->UTF8ToPlatformString(fileMetadata->destFileName);
   std::string srcFileNamePlatform = m_stmtPutGet->UTF8ToPlatformString(fileMetadata->srcFileName);
 
-  FILE *sourceFile = fopen(srcFileNamePlatform.c_str(), "r");
+  FILE *sourceFile;
```
Comment (sfc-gh-igarish) [inline cpp/FileTransferAgent.cpp:620]: do we need to initialize this as null?

File: cpp/FileTransferAgent.cpp [hunk 2]
```diff
@@ -616,12 +617,14 @@ void Snowflake::Client::FileTransferAgent::compressSourceFile(
...
+  sb_fopen(&sourceFile, fileMetadata->srcFileName.c_str(), "r");
   if( !sourceFile ){
     CXX_LOG_ERROR("Failed to open srcFileName %s. Errno: %d", fileMetadata->srcFileName.c_str(), errno);
     throw SnowflakeTransferException(TransferError::FILE_OPEN_ERROR, srcFileNamePlatform.c_str(), -1);
   }
-  FILE *destFile = fopen(stagingFile.c_str(), "w");
+  FILE *destFile;
```
Comment (sfc-gh-igarish) [inline cpp/FileTransferAgent.cpp:626]: do we need to initialize?

File: lib/platform.c [hunk 3]
```diff
@@ -266,12 +266,14 @@ int STDCALL sf_setenv(const char *name, const char *value) {
...
 #ifdef _WIN32
-    return getenv(name);
-#else
-    return getenv(name);
-#endif
+    char* result = NULL;
+    return _dupenv_s(&result, NULL, name) ? NULL : result;
+#   else
+    char* result = getenv(name);
+    return result ? strdup(result) : NULL;
```
Comment (sfc-gh-igarish) [inline lib/platform.c:275]: who will free the buffer of strdup?

File: cpp/FileTransferAgent.cpp [hunk 4]
```diff
@@ -616,12 +617,14 @@ void Snowflake::Client::FileTransferAgent::compressSourceFile(
   stagingFile += m_stmtPutGet->UTF8ToPlatformString(fileMetadata->destFileName);
   std::string srcFileNamePlatform = m_stmtPutGet->UTF8ToPlatformString(fileMetadata->srcFileName);
 
-  FILE *sourceFile = fopen(srcFileNamePlatform.c_str(), "r");
+  FILE *sourceFile = NULL;
+  sb_fopen(&sourceFile, fileMetadata->srcFileName.c_str(), "r");
```
Comment (sfc-gh-ext-simba-hx) [inline cpp/FileTransferAgent.cpp:621]: @sfc-gh-ext-simba-jz  merging issue that breaks put on GCP with unicode characters.
Please double check your changes to make sure no logic changes beside replacement.

**Comment by sfc-gh-ext-simba-jz** [comment]
it is not mature yet, need more test

File: lib/platform.c (hunk 3 above)
Comment (sfc-gh-ext-simba-jz) [reply lib/platform.c]: we will leave getenv and strerror untouched. Need to free them at every invoking. Replace in the next release.

File: cpp/FileTransferAgent.cpp (hunk 2 above)
Comment (sfc-gh-ext-simba-jz) [reply cpp/FileTransferAgent.cpp]: fixed

File: cpp/FileTransferAgent.cpp (hunk 1 above)
Comment (sfc-gh-ext-simba-jz) [reply cpp/FileTransferAgent.cpp]: yes, fixed

File: cpp/FileTransferAgent.cpp (hunk 4 above)
Comment (sfc-gh-ext-simba-jz) [reply cpp/FileTransferAgent.cpp]: fixed. It's a merge mistake.

**Review by sfc-gh-igarish — APPROVED** [review]
LGTM

**Pull request merged**